File: app/src/main/java/ru/MeatGames/roguelike/tomb/screen/BrezenhamView.java
package ru.MeatGames.roguelike.tomb.screen;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Typeface;
import android.view.MotionEvent;
import android.view.View;

import ru.MeatGames.roguelike.tomb.Game;
import ru.MeatGames.roguelike.tomb.Global;
import ru.MeatGames.roguelike.tomb.R;

public class BrezenhamView extends View {
    private Paint white;
    private Paint blue;
    private Paint red;
    private Paint grn;
    private Paint frame;
    private Paint hud;
    private Paint text;
    private int x1;
    private int y1;

    public BrezenhamView(Context c) {
        super(c);
        Global.INSTANCE.setGame((Game) c);
        setFocusable(true);
        setFocusableInTouchMode(true);
        frame = new Paint();
        frame.setColor(getResources().getColor(R.color.mainBackground));
        hud = new Paint();
        hud.setColor(getResources().getColor(R.color.hud));
        white = new Paint();
        white.setColor(getResources().getColor(R.color.white));
        blue = new Paint();
        blue.setColor(getResources().getColor(R.color.lightBlue));
        red = new Paint();
        red.setColor(getResources().getColor(R.color.fredl));
        grn = new Paint();
        grn.setColor(getResources().getColor(R.color.grs));
        text = new Paint(Paint.ANTI_ALIAS_FLAG);
        text.setColor(getResources().getColor(R.color.cell));
        text.setStyle(Style.FILL_AND_STROKE);
        text.setTextScaleX(1);
        text.setTextSize(16);
        text.setTypeface(Typeface.createFromAsset(Global.INSTANCE.getGame().getAssets(), "fonts/Bulgaria_Glorious_Cyr.ttf"));
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawRect(0, 0, 480, 800, frame);
        for (int x = 0; x < Global.INSTANCE.getGame().mw; x++)
            for (int y = 0; y < Global.INSTANCE.getGame().mh; y++) {
                if (Global.INSTANCE.getMap()[x][y].dis) {
                    switch (Global.INSTANCE.getMap()[x][y].o) {
                        case 0:
                            canvas.drawRect(x * 5, 5 + 5 * y, x * 5 + 5, 10 + 5 * y, white);
                            break;
                        case 31:
                        case 32:
                            canvas.drawRect(x * 5, 5 + 5 * y, x * 5 + 5, 10 + 5 * y, hud);
                            break;
                        case 40:
                            canvas.drawRect(x * 5, 5 + 5 * y, x * 5 + 5, 10 + 5 * y, grn);
                            break;
                    }
                    /*if(Global.map[x][y].hasMob())
                        canvas.drawRect(x*5,5+5*y,x*5+5,10+5*y,blue);*/
                }
            }
        canvas.drawRect(Global.INSTANCE.getHero().mx * 5, 5 + Global.INSTANCE.getHero().my * 5, 5 + Global.INSTANCE.getHero().mx * 5, 10 + Global.INSTANCE.getHero().my * 5, red);
        text.setTextAlign(Paint.Align.RIGHT);
        canvas.drawText(getContext().getString(R.string.back_label), 435, 765, text);
        postInvalidate();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_UP:
                x1 = (int) event.getX();
                y1 = (int) event.getY();
                if (y1 > 720 && x1 > 320)
                    Global.INSTANCE.getGame().changeScreen(0);
                break;
        }
        return true;
    }
}

File: app/src/main/java/ru/MeatGames/roguelike/tomb/screen/BrezenhamView.kt
package ru.MeatGames.roguelike.tomb.screen

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.view.MotionEvent
import android.view.View
import ru.MeatGames.roguelike.tomb.Game
import ru.MeatGames.roguelike.tomb.Global
import ru.MeatGames.roguelike.tomb.R
import ru.MeatGames.roguelike.tomb.util.ScreenHelper
import ru.MeatGames.roguelike.tomb.util.UnitConverter
import ru.MeatGames.roguelike.tomb.util.fillFrame
import ru.MeatGames.roguelike.tomb.view.TextButton

// displays explored map
class BrezenhamView(mContext: Context) : View(mContext) {

    val mScreenWidth: Int
    val mScreenHeight: Int

    private val mBackgroundPaint = Paint()
    private val mDoorMarkerPaint = Paint()
    private val mRoomBackgroundPaint = Paint()
    private val mHeroMarkerPaint = Paint()
    private val mExitMarkerPaint = Paint()
    private val mTextPaint: Paint

    private val mBackButton: TextButton

    private val mMarkerSize: Float

    init {
        isFocusable = true
        isFocusableInTouchMode = true

        val screenSize = ScreenHelper.getScreenSize((context as Game).windowManager)
        mScreenWidth = screenSize.x
        mScreenHeight = screenSize.y

        mBackgroundPaint.color = resources.getColor(R.color.mainBackground)
        mDoorMarkerPaint.color = resources.getColor(R.color.hud)
        mRoomBackgroundPaint.color = resources.getColor(R.color.white)
        mHeroMarkerPaint.color = resources.getColor(R.color.fredl)
        mExitMarkerPaint.color = resources.getColor(R.color.grs)

        mTextPaint = ScreenHelper.getDefaultTextPaint(mContext)
        mTextPaint.textSize = UnitConverter.convertSpToPixels(32F, mContext)

        mBackButton = TextButton(context, resources.getString(R.string.back_label))
        mBackButton.mTextPaint.textAlign = Paint.Align.RIGHT
        mBackButton.mDimensions = Rect(mScreenWidth / 3 * 2,
                mScreenHeight - mScreenHeight / 10,
                mScreenWidth,
                mScreenHeight)

        mMarkerSize = (mScreenWidth / Global.game!!.mw).toFloat()
    }

    override fun onDraw(canvas: Canvas) {
        canvas.fillFrame(mScreenWidth, mScreenHeight, mBackgroundPaint)
        for (x in 0..Global.game!!.mw - 1)
            for (y in 0..Global.game!!.mh - 1) {
                if (Global.map!![x][y].mIsDiscovered) {
                    when (Global.map!![x][y].mObjectID) {
                        0 -> canvas.drawRect(x * mMarkerSize,
                                (y + 1) * mMarkerSize,
                                (x + 1) * mMarkerSize,
                                (y + 2) * mMarkerSize,
                                mRoomBackgroundPaint)
                        31, 32 -> canvas.drawRect(x * mMarkerSize,
                                (y + 1) * mMarkerSize,
                                (x + 1) * mMarkerSize,
                                (y + 2) * mMarkerSize,
                                mDoorMarkerPaint)
                        40 -> canvas.drawRect(x * mMarkerSize,
                                (y + 1) * mMarkerSize,
                                (x + 1) * mMarkerSize,
                                (y + 2) * mMarkerSize,
                                mExitMarkerPaint)
                    }
                    /*if(Global.map[x][y].hasMob())
                        canvas.drawRect(x*5,5+5*y,x*5+5,10+5*y,blue);*/
                }
            }
        canvas.drawRect(Global.hero!!.mx * mMarkerSize,
                (Global.hero!!.my + 1) * mMarkerSize,
                (Global.hero!!.mx + 1) * mMarkerSize,
                (Global.hero!!.my + 2) * mMarkerSize,
                mHeroMarkerPaint)
        mBackButton.draw(canvas)
        postInvalidate()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_UP -> {
                val touchX = event.x.toInt()
                val touchY = event.y.toInt()
                if(mBackButton.isPressed(touchX, touchY)) {
                    Global.game!!.changeScreen(0)
                }
            }
        }
        return true
    }
}